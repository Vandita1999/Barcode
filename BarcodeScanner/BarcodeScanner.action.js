import { NavigationService } from '@navis-npm/native-helpers';
import BarcodeScanningService from './BarcodeScanner.service';
import { transformN4TallyDetails } from './BarcodeScanner.utils';
/*eslint-disable-next-line no-unused-vars */
import { ThunkAction } from 'redux-thunk';
/*eslint-disable-next-line no-unused-vars */
import { AppState } from '../common/typedef';
/*eslint-disable-next-line no-unused-vars */
import { Action } from 'redux';
/*eslint-disable-next-line no-unused-vars */
import { tallyDetailsData } from './BarcodeScanner.types';

/**
 *
 * @typedef {tallyDetailsData} tallyDetails
 */
/**
 *
 * @param {string} vinNumber
 * @param {string} operationType
 * @returns {ThunkAction<void,AppState,null,Action<string>>}
 */
const getTallyDetails = (vinNumber, operationType) => (dispatch, getState) => {
  const barcodeScanningService = new BarcodeScanningService();
  return barcodeScanningService
    .getTallyDetails(vinNumber, operationType)
    .then(response => {
      const rawTallyDetails = response.data.content;
      const tallyDetails = transformN4TallyDetails(rawTallyDetails);
      NavigationService.navigate('TallyDetailsScreen', { tallyDetails });
    })
    .catch(error => {
      NavigationService.navigate('RecentTallyScreen');
    });
};
export { getTallyDetails };
