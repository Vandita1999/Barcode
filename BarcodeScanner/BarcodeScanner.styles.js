import { StyleSheet } from 'react-native';
import { moderateScale } from '@navis-npm/native-ui-components';
import * as Colors from '../../assets/colors';
export default StyleSheet.create({
  container: {
    height: '100%',
    width: '100%',
    backgroundColor: Colors.Overlays.background,
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'space-between'
  },
  preview: {
    marginTop: moderateScale(7),
    flex: 7,
    alignSelf: 'center',
    width: '95%',
    height: '100%',
    borderWidth: moderateScale(2),
    borderColor: Colors.White,
    borderRadius: moderateScale(8)
  },
  previewCamera: {
    width: '100%',
    height: '100%',
    alignSelf: 'center',
    overflow: 'hidden',
    borderWidth: moderateScale(2),
    borderRadius: moderateScale(8)
  },
  bottomOverlay: {
    flex: 3,
    marginTop: moderateScale(10),
    height: moderateScale(70),
    width: '100%',
    flexDirection: 'column'
  },
  inputHeader: {
    flex: 4,
    color: Colors.White
  },
  inputHeaderView: {
    height: moderateScale(20),
    marginLeft: '5%'
  },
  inputFieldArea: {
    width: '95%',
    alignSelf: 'center',
    flexDirection: 'row',
    borderWidth: moderateScale(1),
    borderColor: Colors.Black,
    borderRadius: moderateScale(5),
    backgroundColor: Colors.White
  },
  input: {
    textAlign: 'center'
  },
  inputField: {
    flex: 6
  },
  cameraIcon: {
    flex: 1,
    alignSelf: 'center'
  },
  bottomBar: {
    flex: 0,
    marginBottom: 0
  }
});
