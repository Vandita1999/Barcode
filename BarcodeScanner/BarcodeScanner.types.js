/*eslint-disable-next-line no-unused-vars */
import NavigationStackProp from 'react-navigation-stack';
/**
 *
 * @typedef {object} tallyDetailsData
 * @param {string} lotid
 * @param {string} fromPosition
 * @param {string} toPosition
 * @param {string} wiType
 * @param {damageObject} damage
 *
 */
/**
 *
 * @typedef {object} damageObject
 * @param {string} crgdmgSeverity
 * @param {number} crgdmgTypeGkey
 * @param {string} crgdmgtypId
 * @param {number} crgdmgGkey
 * @param {string} crgdmgposX
 * @param {string} crgdmgPosY
 * @param {string} crgdmgRepairRemarks
 * @param {string} crgdmgImg
 */
/**
 * @typedef {object} BarcodeScannerScreenProps
 * @property {function(string):void} getTallyDetails Action to fetch Tally details for VIN number
 * @property {NavigationStackProp} navigation The navigation prop
 * @property {()=> void} onBarCodeRead
 */
/**
 * @typedef {object} BarcodeViewProps
 * @property {function(string):void} getTallyDetails Action to fetch Tally details for VIN number
 * @property {NavigationStackProp} navigation The navigation prop
 * @property {string} barcodeValue Initial state of barcode value
 * @property {boolean} scanBarcode
 * @property {boolean} setScanBarcode
 * @property {string} setBarcodeValue
 * @property {()=> void} onWillFocus
 * @property {()=> void} onWillBlur
 * @property {string} operationType
 */
