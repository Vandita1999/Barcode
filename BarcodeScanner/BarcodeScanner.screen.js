import React from 'react';
import { connect } from 'react-redux';
import BarcodeScannerView from './BarcodeScanner.view';
import { getTallyDetails } from './BarcodeScanner.action';
//eslint-disable-next-line no-unused-vars
import { BarcodeScannerScreenProps } from './BarcodeScanner.types';

/**
 * Component for barcode scanning screen
 * @typedef {React.FC<BarcodeScannerScreenProps>}
 */

export const BarcodeScannerScreen = props => {
  const { getTallyDetails: getTallyDetailsAction, navigation } = props;
  const [barcodeValue, setBarcodeValue] = React.useState('');
  const [scanBarcode, setScanBarcode] = React.useState(true);
  const operationType = navigation.getParam('tallyType');

  const onWillFocus = () => {
    setScanBarcode(true);
  };

  const onWillBlur = () => {
    setBarcodeValue('');
    setScanBarcode(false);
  };

  const onBarCodeRead = barcodeData => {
    if (barcodeValue === '') {
      setScanBarcode(false);
      setBarcodeValue(barcodeData.data);
      getTallyDetailsAction(barcodeData.data, operationType);
    }
  };
  return (
    //@ts-ignore
    <BarcodeScannerView
      navigation={navigation}
      onBarCodeRead={onBarCodeRead}
      barcodeValue={barcodeValue}
      scanBarcode={scanBarcode}
      onWillFocus={onWillFocus}
      onWillBlur={onWillBlur}
      operationType={operationType}
    />
  );
};
export default connect(null, {
  getTallyDetails
})(BarcodeScannerScreen);
