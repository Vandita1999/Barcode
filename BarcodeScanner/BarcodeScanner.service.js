import { apiClient, endpoints } from '../common/roroTally.service';
/*eslint-disable-next-line no-unused-vars */
import { ResponseMessage } from '@navis-npm/nmobile-core';
/*eslint-disable-next-line no-unused-vars */
import { tallyDetailsData } from './BarcodeScanner.types';
/**
 * @typedef {object} StatusResponse
 * @property {tallyDetailsData} content
 * @property {ResponseMessage[]} messages
 */

/**
 * @param {string} vinNumber
 * @param {string} operationType
 * @returns {AxiosPromise<StatusResponse>}
 */
class BarcodeScanningService {
  getTallyDetails(vinNumber, operationType = 'DSCH') {
    return apiClient.instance.get(endpoints.tallyDetails, { params: { cargolotId: vinNumber, wiType: operationType } });
  }
}
export default BarcodeScanningService;
