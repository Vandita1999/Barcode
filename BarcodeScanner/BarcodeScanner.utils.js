/*eslint-disable-next-line no-unused-vars */
import { tallyDetailsData } from './BarcodeScanner.types';
/**
 *
 * @param {tallyDetailsData} rawTallyDetails
 * @returns
 */
export const transformN4TallyDetails = rawTallyDetails => {
  let tallyDetails = {};
  tallyDetails.vin = rawTallyDetails.lotId;
  tallyDetails.origin = rawTallyDetails.fromPosition;
  tallyDetails.destination = rawTallyDetails.toPosition;
  tallyDetails.operation = rawTallyDetails.wiType;

  let damages = rawTallyDetails.damage.map(damage => {
    let transformedDamage = {};
    transformedDamage.positionX = damage.crgdmgposX;
    transformedDamage.positionY = damage.crgdmgPosY;
    transformedDamage.type = damage.crgdmgtypId;
    transformedDamage.typeGkey = damage.crgdmgTypeGkey;
    transformedDamage.severity = damage.crgdmgSeverity.slice(-7).slice(1, 6);
    transformedDamage.notes = damage.crgdmgRepairRemarks;
    transformedDamage.imageData = damage.crgdmgImg;

    return transformedDamage;
  });
  tallyDetails.damages = damages;
  return tallyDetails;
};
