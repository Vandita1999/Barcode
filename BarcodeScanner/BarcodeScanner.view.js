import React from 'react';
import { View, TextInput, Text } from 'react-native';
import { NavigationEvents } from 'react-navigation';
import { RNCamera } from 'react-native-camera';
import barcodeStyles from './BarcodeScanner.styles';
import { strings } from '../../assets/strings/strings';
import { Icon } from '@navis-npm/native-ui-components';
import BottomBar from '../common/BottomBar/BottomBar.view';
/*eslint-disable-next-line no-unused-vars */
import { BarcodeViewProps } from './BarcodeScanner.types';

/**
 *
 * @typedef {React.FC<BarcodeViewProps>}
 */
const BarcodeScannerView = ({ navigation, onBarCodeRead, barcodeValue, scanBarcode, onWillFocus, onWillBlur, operationType }) => {
  const refRNCamera = React.useRef(null);
  return (
    <View style={barcodeStyles.container}>
      <NavigationEvents onWillFocus={onWillFocus} onWillBlur={onWillBlur} />
      {scanBarcode && (
        <View style={barcodeStyles.preview}>
          <RNCamera ref={refRNCamera} type={RNCamera.Constants.Type.back} captureAudio={false} style={barcodeStyles.previewCamera} onBarCodeRead={onBarCodeRead} />
        </View>
      )}
      <View style={barcodeStyles.bottomOverlay}>
        <View style={barcodeStyles.inputHeaderView}>
          <Text style={barcodeStyles.inputHeader}>{strings.en.labels.vin}</Text>
        </View>
        <View style={barcodeStyles.inputFieldArea}>
          <View style={barcodeStyles.inputField}>
            <TextInput style={barcodeStyles.input} value={barcodeValue} />
          </View>
          <View style={barcodeStyles.cameraIcon}>
            <Icon name={strings.en.icons.camera} width={20} height={18} />
          </View>
        </View>
      </View>
      <View style={barcodeStyles.bottomBar}>
        <BottomBar
          screenTitle={operationType === strings.en.labels.load ? strings.en.placeholders.tallyLoad : strings.en.placeholders.tallyDischarge}
          navigation={navigation}
          hasBackIcon={true}
        />
      </View>
    </View>
  );
};

export default BarcodeScannerView;
