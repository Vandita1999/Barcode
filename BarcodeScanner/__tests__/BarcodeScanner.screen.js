import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { BarcodeScannerScreen } from '../BarcodeScanner.screen';

const setUpBarcodeScannerScreen = newProps => {
  let props = {
    getTallyDetails: jest.fn(),
    navigation: {
      getParam: jest.fn().mockReturnValue('DSCH')
    }
  };
  props = { ...props, ...newProps };
  // @ts-ignore
  const component = shallow(<BarcodeScannerScreen {...props} />);
  return { props, component };
};
describe('Barcode Scanner screen.', () => {
  beforeEach(jest.clearAllMocks);
  it('Should match snaphot', () => {
    const { component } = setUpBarcodeScannerScreen();
    expect(toJson(component)).toMatchSnapshot();
  });
  it('Should set scanBarcode as false on Wil Focus', () => {
    const { component } = setUpBarcodeScannerScreen();
    expect(component.props().scanBarcode).toBe(true);
    component.props().onWillFocus();
    expect(component.props().scanBarcode).toBe(true);
  });
  it('Should set scanBarcode as false on Will Blur', () => {
    const { component } = setUpBarcodeScannerScreen();
    expect(component.props().scanBarcode).toBe(true);
    component.props().onWillBlur();
    expect(component.props().scanBarcode).toBe(false);
  });
  it('Should set barcodeValue when `onBarCodeRead` is called', () => {
    const mockGetTallyDetails = jest.fn();
    const { component } = setUpBarcodeScannerScreen({ getTallyDetails: mockGetTallyDetails });
    expect(component.props().barcodeValue).toBe('');
    component.props().onBarCodeRead({ data: '123' });
    expect(component.props().barcodeValue).toBe('123');
    expect(mockGetTallyDetails).toBeCalledWith('123', 'DSCH');
    component.props().onBarCodeRead({ data: '345' }); //Should not change the barcode value now
    expect(component.props().barcodeValue).toBe('123');
  });
});
