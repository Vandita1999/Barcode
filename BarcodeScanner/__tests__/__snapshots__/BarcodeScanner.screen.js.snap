// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barcode Scanner screen. Should match snaphot 1`] = `
<BarcodeScannerView
  barcodeValue=""
  navigation={
    Object {
      "getParam": [MockFunction] {
        "calls": Array [
          Array [
            "tallyType",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "DSCH",
          },
        ],
      },
    }
  }
  onBarCodeRead={[Function]}
  onWillBlur={[Function]}
  onWillFocus={[Function]}
  operationType="DSCH"
  scanBarcode={true}
/>
`;
